runners/core-java/src/main/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvoker.java

```diff
@@ -150,10 +150,19 @@ public Timer timer(String timerId) {
                 "Access to timers not supported in Splittable DoFn");
           }
         });
-
+    // TODO: verify that if there was a failed tryClaim() call, then cont.shouldResume() is false.
+    // Currently we can't verify this because there are no hooks into tryClaim().
+    RestrictionT residual = processContext.extractCheckpoint();
+    if (cont.shouldResume()) {
+      if (residual == null) {
+        residual = tracker.checkpoint();
+      } else {
+        // This is ok - checkpoint taken by processContext is equivalent to the one we could have
```

Comment:
We should describe why the `tracker.checkpoint()` call above is required (and will return a checkpoint even if `processContext` didn't).